man1/javachimes.1
.\" Copyright (C) 2003-24 Kevin Boone 
.\" Permission is granted to any individual or institution to use, copy, or
.\" redistribute this software so long as all of the original files are
.\" included, and that this copyright notice is retained.
.\"
.TH epub2txt 1 "June 2024"
.SH NAME
javachimes \- a wind chimes simulator 
.SH SYNOPSIS
.B javachimes 
.RB [options]
.PP

.SH DESCRIPTION
\fIjavachimes\fR simulates the  
sound of a mechanical wind-chime using MIDI notes generated
using the Java sound API. By default, it will use the 
tone generator built into the Java virtual machine, but it can be
configured to use an external MIDI device if you have one. 
It supports soundbanks in SoundFont 2 format, or it can use the 
JVM's built in instrument patches.


.SH ALGORITHM
\fIjavachimes\fR 
is controlled by two main parameters: 
`windspeed' (w) and (lacking a better term) `gustyness' (g). The
average rate of tones is proportional to the windspeed parameter, while
`gustyness' controls the variability in this rate.  If g=0, then 
the rate of generation of
tones will be approximately constant.  Technically, the rate of generation (in
seconds) is randomly selected from a normal distribution whose mean is w/10,
and whose standard deviation is g/10. 

A low windspeed (e.g., < 10), coupled with a high gustyness (e.g., > 20) gives
a pattern of intense bursts of activity separated by long pauses, which is
quite typical or real wind-chimes. However, the long periods of silence
might make it difficult to tell whether the program is working.

.SH "OPTIONS"
.TP
.BI -a,\-\-channel\ (number)
Sets the MIDI output channel, in the range 1-16. This is not particularly
relevant for soundboard-based synthesizers, as they typically respond on all
channels. However, it may well be important for external sound generators. In
all cases, bear in mind the General MIDI specifications call for channel 10 to
be reserved for percussion effects. The default value is 1.
.LP
.TP
.BI -b,\-\-basenote\ (1-128) 
The base (root) note of the scale, when using one of the built-in scales.
This is specified as a MIDI note number, where middle C is `60' and adding one
increases the pitch by a semitone (so the C two octaves above middle C is 84
-- 60 + 12 + 12). This parameter is ignored when the --scale switch is in use.
The default value is 85.
.LP
.TP
.BI -c,\-\-scale\ (n1,n2,n3...) 
Specifies the exact set of notes to use, as MIDI note numbers separated
by commas.
.LP
.TP
.BI -d,\-\-onduration\ (seconds) 
Set the note-on duration. The default
is two seconds. Longer values sound more realistic, especially at lower
pitches, but at the expense of increased memory and CPU usage.
.LP
.TP
.BI -e,\-\-device\ (number) 
Specify the MIDI device. The default is to use the first one known to the
JVM, which is usual the internal sound generator.
.LP
.TP
.BI -f,\-\-soundfont\ (file) 
Java usually supports soundbanks in SoundFont 2 format; files usually have
names ending in `.sf2`. Some JVMs may support other formats. See 'resources'
section below for information about configuring memory, if necessary.
If this option is not given, JavaChimes will use default instrument sounds.
.LP
.TP
.BI -g,\-\-gustiness\ (0-100) 
Sets the `gustyness' of the wind; this affects the variability in the
time delays between chimes. See `algorithm' below. The default value is
30.
.LP
.TP
.BI -l,\-\-listdevs
Lists the know MIDI devices with their numbers. The number can then be used
as a parameter to the `--device` switch to select from a range of installed
devices or drivers.
.LP
.TP
.BI -n,\-\-numchimes\ (1-10) 
Select the number of chimes to use. `1' is allowed, but not very interesting.
The default is 5. 
.LP
.TP
.BI -p,\-\-program\ (1-128) 
Selects the program (voice, instrument) to use. See notes below. Default
is 14: usually xylophone or bell.
.LP
.TP
.BI -s,\-\-seconds\ (number) 
Specifies the number of seconds for which to play. The default is to
play forever, until the program is forcibly stopped by control-C or 
whatever.
.LP
.TP
.BI -t,\-\-scaletype\ (0-11) 
Sets which of the pre-defined scales to use. The default is `8', a 
pentatonic scale. This switch is over-ridden by `--scale'. See
\fIscales\fR below.
.LP
.TP
.BI -w,\-\-windspeed\ (1-100) 
Sets the wind-speed, which affects the rate production of sound events.
Default is 20.
.LP
.TP
.BI -v,\-\-version
Fooo
.LP
.TP
.BI -h,\-\-help
Shows command-line usage.
.LP

.SH "SCALES"

JavaChimes supports a number of pre-defined scales, or an exact scale can be
specified on the command line. The default is to use a pentatonic scale.
Other scales are:

0: major (7 notes) 

1: natural minor (7 notes)  

2: harmonic minor (7 notes)  

3: melodic minor (7 notes)  

4: enigmatic (7 notes)  

5: chromatic (11 notes)  

6: arpeggio (3 notes)  

7: dominant (`barbershop') seventh (4 notes)  

8: pentatonic (5 notes) (this is the default)  

9: Greek pentatonic (5 notes)  

10: Gregorian (7 notes)  

11: Hijazi (7 notes)  

The notes of the scale are determined by the base note and 
number of chimes, as well as the scale itself, 

To use an exact scale, give the \fI--scale\fR argument followed by
a list of MIDI note numbers, separated by commas.

.SH ENVIRONMENT 

The environment variable \fIJAVA_OPTS\fR is passed directly to the
java command line. This can be useful to specify a heap size limit,
for example.




.SH AUTHOR AND LEGAL
\fIjavachimes\fR is maintained by Kevin Boone, and is open source under the
terms of the GNU Public Licence, version 3.0. There is no warranty of any kind.


.\" end of file


src/main/java/me/kevinboone/javachimes/JavaChimes.java
package me.kevinboone.javachimes;
import javax.sound.midi.*;
import java.util.*;


/**
*/
public class JavaChimes extends Thread 
{
/*
Define the scales available, and constants to allow
them to be selected. Note that the first note is always
`0', that is, the same as the base note. The values in the
arrays are semitones above the base note
*/
public static int[] majorScale = {0, 2, 4, 5, 7, 9, 11};
public static int[] minorScale = {0, 2, 3, 5, 7, 8, 10};
public static int[] harmonicScale = {0, 2, 3, 5, 7, 11};
public static int[] melodicScale = {0, 2, 3, 5, 7, 9, 11};
public static int[] enigmaticScale = {0, 1, 4, 6, 8, 10, 11};
public static int[] chromaticScale = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11};
public static int[] arpeggioScale = {0, 4, 7};
public static int[] domseventhScale = {0, 4, 7, 11};
public static int[] pentatonicScale = {0, 2, 5, 7, 9};
public static int[] greekpentatonicScale = {0, 1, 5, 7, 8};
public static int[] gregorianScale = {0, 2, 3, 5, 7, 9, 10};
public static int[] hijaziScale = {0, 1, 4, 5, 6, 8, 11};


public static final int SCALETYPE_MAJOR = 0;
public static final int SCALETYPE_MINOR = 1;
public static final int SCALETYPE_HARMONIC = 2;
public static final int SCALETYPE_MELODIC = 3;
public static final int SCALETYPE_ENIGMATIC = 4;
public static final int SCALETYPE_CHROMATIC = 5;
public static final int SCALETYPE_ARPEGGIO = 6;
public static final int SCALETYPE_DOMSEVENTH = 7;
public static final int SCALETYPE_PENTATONIC = 8;
public static final int SCALETYPE_GREEKPENTATONIC = 9;
public static final int SCALETYPE_GREGORIAN = 10;
public static final int SCALETYPE_HIJAZI = 11;
public static final int SCALETYPE_USER = -1;

/**
A reference to a MIDI synth object
*/
protected Synthesizer synthesizer;

protected int program = 15 /* tub. bell */ - 1;
protected boolean quit = false;
/**
MIDI channel assigned during the `open()' call
*/
protected MidiChannel channel = null;
protected int noteOnDurationMillis = 2000;
protected int windSpeed = 20;
protected int gustyness = 30;
protected int baseNote = 85; 
protected int numChimes = 5;
protected int scaleType = SCALETYPE_PENTATONIC;
protected int device = -1;
protected int channelNumber = 0;
protected Random random = new Random();
protected int[] scale;
protected String soundbankFile = null;

/**
Constructor readies the synth device, etc
*/
public JavaChimes() throws JavaChimesException
  {
  // Make a default scale, in case the client calls start()
  //  without calling any of the property-setting methods
  makeScale();
  }

/**
Closes the MIDI system and frees resources
*/
protected void close() 
  {
  if (synthesizer != null) 
    synthesizer.close();
  synthesizer = null;
  channel = null;
  }

protected void checkRange (final int x, final int min, 
    final int max, final String method)
  throws JavaChimesIllegalArgumentException
  {
  if (x < min || x > max) 
    throw new JavaChimesIllegalArgumentException 
      ("Argument out of range (" + min + "-" + max + ") in " + method + "()"); 
  }

public void setSoundbankFile (String soundbankFile)
  {
  this.soundbankFile = soundbankFile;
  }

/**
Sets the (one-based) MIDI program number for output
*/
public void setProgram (final int _program)
  {
  checkRange (_program, 1, 128, "setProgram");
  program = _program;
  if (channel != null)
    channel.programChange(_program - 1);
  }

public int getProgram ()
  {
  return program + 1;
  }

/**
Sets the (one-based) MIDI channel number for output
*/
public void setChannel (final int _channelNumber)
  {
  checkRange (_channelNumber, 1, 16, "setChannel");
  channelNumber = _channelNumber  - 1;
  }

public int getChannel ()
  {
  return channelNumber + 1;
  }

/**
Sets the base note for the scale, 0-127 
*/
public void setBaseNote (final int _baseNote)
  {
  checkRange (_baseNote, 0, 127, "setBaseNote");
  baseNote = _baseNote;
  makeScale();
  }

public int getBaseNote ()
  {
  return baseNote;
  }

/**
Sets the scale type to the specified constant 
*/
public void setScaleType (final int _scaleType)
  {
  checkRange (_scaleType, 0, 11, "setScaleType");
  scaleType = _scaleType;
  makeScale();
  }

public int getScaleType ()
  {
  return scaleType;
  }

/**
Sets the scale to the specified note intervals 
*/
public void setScale (final int[] _scale)
  {
  scale = _scale;
  scaleType = SCALETYPE_USER;
  }

public int[] getScale ()
  {
  return scale;
  }

/**
Sets the device number; -1 = system default 
*/
public void setDevice (final int _device)
  {
  device = _device;
  }

public int getDevice ()
  {
  return device;
  }

/**
Sets the number of chimes. Theoretically this depends on the
scale in use. 
*/
public void setNumChimes (final int _numChimes)
  {
  checkRange (_numChimes, 1, 128, "setNumChimes");
  numChimes = _numChimes;
  makeScale();
  }

public int getNumChimes ()
  {
  return numChimes;
  }

/**
Sets the note-on duration to the specified number of seconds.  
*/
public void setNoteOnDuration (final int _noteOnDuration)
  {
  checkRange (_noteOnDuration, 1, 0x7FFFFFFF, "setNoteOnDuration");
  noteOnDurationMillis = _noteOnDuration * 1000;
  }

public int getNoteOnDuration ()
  {
  return noteOnDurationMillis / 1000;
  }

/**
Sets the windspeed
*/
public void setWindSpeed (final int _windSpeed)
  {
  checkRange (_windSpeed, 1, 100, "setWindSpeed");
  windSpeed = _windSpeed;
  }

public int getWindSpeed ()
  {
  return windSpeed;
  }

/**
Sets the gustyness 
*/
public void setGustyness (final int _gustyness)
  {
  checkRange (_gustyness, 0, 100, "setGustyness");
  gustyness = _gustyness;
  }

public int getGustyness ()
  {
  return gustyness;
  }


/**
Causes the sound generation to stop after the next note has been played,
and the MIDI system to be closed.
*/
public void signal()
  {
  quit = true;
  }

/**
Causes the sound generation to stop after the next note has been played,
and the MIDI system to be closed. This method does not return to the 
caller until these operations have been effected
*/
public void signalAndWait()
  {
  quit = true;
  while (synthesizer != null){};
  }


/**
Returns an array of names of names of available MIDI devices.
*/
public static String[] getDeviceNames () throws JavaChimesException
  {
  MidiDevice.Info info[] = MidiSystem.getMidiDeviceInfo();
  Vector v = new Vector();
  for (int i = 0; i < info.length; i++)
    v.add(info[i].getName());
  String[] t = new String[v.size()];;
  for (int i = 0; i < info.length; i++)
    t[i] = (String)v.elementAt(i);
  return t;
  }

/**
Opens the MIDI system and retrieves a MidiChannel instance for subsequent
output operations
*/
public void open () throws JavaChimesException
  {
  try
    {
    if (device == -1)
      {
      synthesizer = MidiSystem.getSynthesizer();
      }
    else
      {
      MidiDevice.Info info[] = MidiSystem.getMidiDeviceInfo();
      if ((device < 0) || (device >= info.length))
        throw new JavaChimesException ("Device number out of range");
      synthesizer = (Synthesizer)MidiSystem.getMidiDevice(info[device]);
      }
    synthesizer.open();
            //sequencer = MidiSystem.getSequencer();
            //sequence = new Sequence(Sequence.PPQ, 10);
    }
  catch (MidiUnavailableException e)
    {
    throw new JavaChimesException ("Can't get access to a synth device");
    }
  catch (ClassCastException e)
    {
    throw new JavaChimesException ("Selected MIDI device is not a synthesizer");
    }

  if (soundbankFile == null)
    {
    Soundbank sb = synthesizer.getDefaultSoundbank();
    if (sb != null) 
      {
      Instrument[] instruments = synthesizer.getDefaultSoundbank().getInstruments();
      synthesizer.loadInstrument(instruments[0]);
      }
    }
  else
    {
    try
      {
      MidiSystem.getSequencer(false); // Workaround for JVM bug
      synthesizer.unloadAllInstruments (synthesizer.getDefaultSoundbank());
      Soundbank sb = MidiSystem.getSoundbank 
           (new java.io.File (soundbankFile));
      synthesizer.loadAllInstruments (sb);
      }
    catch (Exception e)
      { 
      //e.printStackTrace();
      throw new JavaChimesException ("Can't load soundbank file: " 
        + e.toString());
      }
    }

  MidiChannel midiChannels[] = synthesizer.getChannels();
  channel = midiChannels[channelNumber];
  }

/**
Recalculates the notes in the scale array. Client classes can call
this, but it is called automatically from setBaseNote(), etc.
*/
public void makeScale()
  {
  if (scaleType == SCALETYPE_USER) return;
  int currentBaseNote = baseNote;
  int[] selectedScale;
  switch (scaleType)
    {
    case SCALETYPE_MAJOR:
      selectedScale = majorScale;
      break;
    case SCALETYPE_MINOR:
      selectedScale = minorScale;
      break;
    case SCALETYPE_HARMONIC:
      selectedScale = harmonicScale;
      break;
    case SCALETYPE_MELODIC:
      selectedScale = melodicScale;
      break;
    case SCALETYPE_ENIGMATIC:
      selectedScale = enigmaticScale;
      break;
    case SCALETYPE_CHROMATIC:
      selectedScale = chromaticScale;
      break;
    case SCALETYPE_DOMSEVENTH:
      selectedScale = domseventhScale;
      break;
    case SCALETYPE_ARPEGGIO:
      selectedScale = arpeggioScale;
      break;
    case SCALETYPE_GREEKPENTATONIC:
      selectedScale = greekpentatonicScale;
      break;
    case SCALETYPE_GREGORIAN:
      selectedScale = gregorianScale;
      break;
    case SCALETYPE_HIJAZI:
      selectedScale = hijaziScale;
      break;
    default:
      selectedScale = pentatonicScale;
    }
  scale = new int[numChimes];
  int ssIndex = 0;
  for (int i = 0; i < numChimes; i++)
    {
    scale[i] = currentBaseNote + selectedScale[ssIndex++];
    if (ssIndex >= selectedScale.length)
      {
      ssIndex = 0;
      int lastNote = selectedScale[selectedScale.length - 1];
      if (lastNote > 36)
        currentBaseNote += 48;
      else if (lastNote > 24)
        currentBaseNote += 36; 
      else if (lastNote > 12)
        currentBaseNote +=24; 
      else
        currentBaseNote +=12; 
      } 
    
    // Shouldn't happen, but protect against generating
    //  notes > 127
    if (currentBaseNote > 127) currentBaseNote -= 12;
    }
  }

/**
Do everything, until something calls stop()
*/
public void run() throws JavaChimesException 
  {
  Vector noteList = new Vector();
  quit = false;
  open();
  channel.programChange(program);
  int notesInScale = scale.length;

  while (!quit)
    {
    double eventsPerSec = random.nextGaussian() * gustyness/10.0 
      + windSpeed/10.0;
    if (eventsPerSec <= 0.0) eventsPerSec = 1;
    double maxEventsPerSec = windSpeed/10.0 + 3.0 * gustyness/10.0;
    int delayMillis = (int)((1.0 / eventsPerSec) * 1000.0);

    try 
      { 
      Thread.sleep(delayMillis); 
      } catch (Exception e){}
    long timeNow = System.currentTimeMillis();

    int velocity = (int)(eventsPerSec / maxEventsPerSec * 127.0);
    if (velocity < 0) velocity = 0;
    if (velocity > 127) velocity = 127;

    // On each pass, generate a new note, and check which existing notes
    //  are sounding and need to be cleared
    int kNum = scale[random.nextInt(notesInScale)]; 
    channel.noteOn(kNum, velocity); 
    NoteInfo ni = new NoteInfo (kNum, timeNow);
    noteList.add (ni);
    
    Iterator it = noteList.iterator();
    while (it.hasNext())
      {
      ni = (NoteInfo)it.next();
      if (timeNow - ni.startTime > noteOnDurationMillis) // FRIG 
        {
        channel.noteOff(ni.kNum);
        noteList.remove(ni);
        it = noteList.iterator();
        }
      }
    }
  close();
  }

}




